### PR TITLE
Timeline drawer 1/3

### PR DESCRIPTION
**Pull request by nkanazawa1989**
### Summary
This PR adds a part of timeline drawer that is a component of scheduled circuit framework #4555 .
This PR is the first of three chained PRs.

### Details and comments
The drawer framewrok inherits from the v2 pulse drawer #4643 which is currently under review.
Details are written in the module docstring.

This PR contains common data type definitions and drawing objects (IRs) with program parser (BitEvents).

### CODE REVIEW

File: qiskit/visualization/timeline/drawing_objects.py
Comment (taalexander) [inline qiskit/visualization/timeline/drawing_objects.py:55]:
```suggestion
Drawing objects for the timeline drawer.

Drawing objects play two important roles:
    - Enabling unittests of the visualization module.
    - Removing program parser from each plotter interface. We can now easily add new plotters.

This module is based on the structure of matplotlib as it is the primary plotter
of the timeline drawer. However this interface is agnostic to the actual plotter implementation.

Design concept
~~~~~~~~~~~~~~
When we think about dynamically updating drawing objects, it will be most efficient to
update only the changed properties of drawings rather than regenerating entirely from scratch.
Thus the core drawing function generates all possible drawings in the beginning and
then updates the visibility and the offset coordinate of each item according to
the end-user request.

Data key
~~~~~~~~
In the abstract class ``ElementaryData`` common properties that represent a drawing object are
specified. In addition, drawing objects have the `data_key` property that returns a
unique hash of the object for comparison. This property should be defined in each sub-class by
considering necessary properties to identify that object, i.e. `visible` should not
be a part of the key, because any change on this property just sets the visibility of
the same drawing object.

To support other plotters than matplotlib, drawing objects should be
universal and designed without strong dependency on modules in `matplotlib`.
This means drawing objects that represent primitive geometries are preferred.
It should be noted that there will be no unittest for a plotter interface, which takes
drawing objects and output an image data, we should avoid adding a complicated data structure
that has a context of the scheduled circuit program.

Usually a drawing object is associated to the specific bit. `BitLinkData` is the
exception of this framework because it takes multiple bits to connect.
Position of the link is dynamically updated according to the user preference of bit to show.
While other objects are general format to represent specific shapes, the `BitLinkData` is
only used by the bit links and thus has no `data_type` input.
```

File: qiskit/visualization/timeline/events.py
Comment (taalexander) [inline qiskit/visualization/timeline/events.py:33]:
```suggestion
Bit event manager for scheduled circuits.

This module provides a `BitEvents` class that manages a series of instructions for a
specific circuit bit. Bit-wise filtering of the circuit program makes the arrangement of bits
easier in the core drawer function. The `BitEvents` class is expected to be called
by other programs (not by end-users).

The `BitEvents` class instance is created with the class method ``load_program``:
    ```python
    event = BitEvents.load_program(sched_circuit, qregs[0])
    ```

Loaded circuit instructions are saved as ``ScheduledGate``, which is a collection of instruction,
associated time, and bits. All gate instructions are returned by the `gates` method.
Instruction types specified in `BitEvents._non_gates` are not considered as gates.
If the instruction is associated with multiple bits and the target bit of the class instance is
the primary bit of the instruction, the instance also generates a ``GateLink`` object
that shows the relationship between bits during multi-bit gates.
```

File: qiskit/visualization/timeline/events.py
Comment (taalexander) [inline qiskit/visualization/timeline/events.py:24]: This isn't latex, should be markdown.

File: qiskit/visualization/timeline/types.py
Comment (taalexander) [inline qiskit/visualization/timeline/types.py:31]: Could this documentation be attached to the `NamedTuple`s https://stackoverflow.com/a/20388499?

File: qiskit/visualization/timeline/types.py
Comment (taalexander) [inline qiskit/visualization/timeline/types.py:34]: I really like your use of types.

File: test/python/visualization/timeline/test_events.py
Comment (taalexander) [inline test/python/visualization/timeline/test_events.py:1]: It seems like `load_program` isn't being tested?

File: qiskit/visualization/timeline/types.py
Comment (nkanazawa1989) [reply qiskit/visualization/timeline/types.py]: Thanks! I didn't notice this is supported by py3.5. Now it looks cleaner.

File: test/python/visualization/timeline/test_events.py
Comment (nkanazawa1989) [reply test/python/visualization/timeline/test_events.py:1]: Right. We don't have #4555 merged and there is no input object for `load_program`. This will be added once it is merged.

File: qiskit/visualization/timeline/__init__.py
Comment (lcapelluto) [inline qiskit/visualization/timeline/__init__.py:2]:
```suggestion
```
Have since been removed everywhere

File: qiskit/visualization/timeline/drawing_objects.py
Comment (lcapelluto) [inline qiskit/visualization/timeline/drawing_objects.py:2]:
```suggestion
```

File: qiskit/visualization/timeline/drawing_objects.py
Comment (lcapelluto) [inline qiskit/visualization/timeline/drawing_objects.py:123]: Other backends should work, too, right? Should this not be
```suggestion
            styles: Style keyword args of the object. This conforms to the plotter interface.
```
?

File: qiskit/visualization/timeline/drawing_objects.py
Comment (lcapelluto) [inline qiskit/visualization/timeline/drawing_objects.py:249]:
```suggestion
    Note this object takes multiple bits and dedicates them to the bit link.
```

File: qiskit/visualization/timeline/events.py
Comment (lcapelluto) [inline qiskit/visualization/timeline/events.py:2]:
```suggestion
```

Comment (lcapelluto) [reply qiskit/visualization/timeline/events.py:2]: same for other files

File: qiskit/visualization/timeline/events.py
Comment (lcapelluto) [inline qiskit/visualization/timeline/events.py:105]:
```suggestion
        return not any(not isinstance(inst, self._non_gates) for inst in self.instructions)
```

File: qiskit/visualization/timeline/events.py
Comment (lcapelluto) [inline qiskit/visualization/timeline/events.py:101]:
```suggestion
        """Return False iff there is any gate associated with this bit."""
```

**Review by lcapelluto — COMMENTED** [review]
Looks good!

File: qiskit/visualization/timeline/drawing_objects.py
Comment (lcapelluto) [inline qiskit/visualization/timeline/drawing_objects.py:55]: This text already exists in `qiskit/visualization/pulse_v2/drawing_objects.py` Could you reference it instead of repeating it?

**Review by lcapelluto — CHANGES_REQUESTED** [review]

File: qiskit/visualization/timeline/drawing_objects.py
Comment (nkanazawa1989) [reply qiskit/visualization/timeline/drawing_objects.py]: I updated a text. Eventually I want to make some base plotter module to provide a skeleton and write everything there.

File: qiskit/visualization/timeline/drawing_objects.py
Comment (nkanazawa1989) [reply qiskit/visualization/timeline/drawing_objects.py:149]: No, this is literary conforms to the matplotlib format, i.e. kwarg name is based on matplotlib.

Comment (lcapelluto) [reply qiskit/visualization/timeline/drawing_objects.py:149]: So, you would need a new `LineData` object to support a different plotter? Or is there somewhere that styles are converted to matplotlib kwargs to the plotter kwargs? I didn't see that anywhere

Comment (nkanazawa1989) [reply qiskit/visualization/timeline/drawing_objects.py:149]: Exactly. As the initial step I'm going to introduce matplotlib-based plotter, but in future we may introduce some other plotters like bokeh. Then we need to write some kwarg converter for that plotter. At this moment I don't plan to make some general stylesheet framework.

**Review by lcapelluto — APPROVED** [review]

**Review by lcapelluto — APPROVED** [review]

File: qiskit/visualization/timeline/drawing_objects.py
Comment (taalexander) [inline qiskit/visualization/timeline/drawing_objects.py:76]:
```suggestion
import numpy as np

from qiskit import circuit
```

File: qiskit/visualization/timeline/types.py
Comment (taalexander) [inline qiskit/visualization/timeline/types.py:132]:
```suggestion
    MPL = 'mpl'
```

**Review by taalexander — APPROVED** [review]
Approving with two non-blocking changes suggested.

**Review by lcapelluto — APPROVED** [review]

**Pull request merged**
